src/core/heila-core.js

```diff
@@ -0,0 +1,311 @@
+          return user.uuid;
+        })
+        //console.log(heilaIds)
+        return knex('heilas')
+          .select('heilas.*')
+          .whereIn('uuid', heilaIds)
+          .then(heilaRows => {
+            //console.log(heilaRows);
+            if (_.isEmpty(heilaRows)) {
+              return [];
+            }
+            const unfilteredHeilalist = _heilaRowsToObjectList(_mergeUserHeilaRows(userRows, heilaRows));
+            //console.log('unfilteredHeilaList:')
+            //console.log(unfilteredHeilalist)
+            const myType = unfilteredHeilalist.filter(h => h.id == userId)[0].bio_looking_for_type_id || -1;
+            //console.log('my_type');
+            //console.log(myType);
+            //console.log('unfilteredList:');
+            //console.log(unfilteredHeilalist);
+
+            return knex('matches')
+              .select('matches.*')
+              .where({ 'userId1': userId })
+              .orWhere({ 'userId2': userId })
+              .then(previousMatches => {
+                //console.log('all previous matches:');
+                //console.log(previousMatches);
+                const filterOutIds = previousMatches.map(match => match.userId1 === userId ? match.userId2 : match.userId1);
+                //console.log('filterOudIds');
+                //console.log(filterOutIds);
+
+                const filtered = unfilteredHeilalist.filter(elem => {
+                  return filterOutIds.indexOf(parseInt(elem.id)) === -1;
+                });
+
+                const sameTypes = [];
+                const otherTypes = [];
+
+                filtered.forEach(heila => {
+                  if (heila.bio_looking_for_type_id === myType) {
+                    sameTypes.push(heila);
+                  } else {
+                    otherTypes.push(heila);
+                  }
+                });
+                return sameTypes.concat(otherTypes);
+              })
+          })
+      })
+  })
+}
+
+function getHeilaByUserId(userId) {
+  return knex('users')
+    .select('users.*')
+    .where({ id: userId })
+    .then(userRows => {
+      if (_.isEmpty(userRows)) {
+        return [];
+      }
+      const heilaIds = userRows.map(user => {
+        return user.uuid;
+      })
+      return knex('heilas')
+        .select('heilas.*')
+        .whereIn('uuid', heilaIds)
+        .then(heilaRows => {
+          if (_.isEmpty(heilaRows)) {
+            return [];
+          }
+          return _heilaRowsToObjectList(_mergeUserHeilaRows(userRows, heilaRows))[0];
+        })
+    })
+
+}
+
+function _mergeUserHeilaRows(userRows, heilaRows) {
+  // mergetään user-profiilit ja heila-profiilit
+  // lopputuloksena palautetaan lista, jossa
+  // sekä userin tiedot että heilan tiedot samassa
+  // objektissa
+  userRows.forEach(user => {
+    for (let i = 0; i < heilaRows.length; i += 1) {
+      if (user.uuid === heilaRows[i].uuid) {
+        user = Object.assign(user, heilaRows[i]);
```

Comment:
To avoid unexpected mutations it is a good practice to always use empty object as first element of Object.assign. Object.assign mutates the first element.

`Object.assign({}, user, heilaRows[i]);`